dodiscover/ci/clf_test.py

```diff
@@ -0,0 +1,359 @@
+from typing import Callable, Optional, Set, Tuple, Union
+
+import numpy as np
+import pandas as pd
+import scipy.special
+import sklearn
+import sklearn.metrics
+from numpy.typing import NDArray
+from sklearn.neighbors import NearestNeighbors
+from sklearn.utils import check_random_state
+
+from dodiscover.typing import Column
+
+from .base import BaseConditionalIndependenceTest
+
+
+class ClassifierCITest(BaseConditionalIndependenceTest):
+    def __init__(
+        self,
+        clf: sklearn.base.BaseEstimator,
+        metric: Callable = sklearn.metrics.accuracy_score,
+        bootstrap: bool = False,
+        n_iter: int = 20,
+        correct_bias: bool = True,
+        threshold: float = 0.03,
+        test_size: Union[int, float] = 0.3,
+        random_state: Optional[int] = None,
```

Comment:
The random state should be set outside of the class objects. If a user wants to fix the seed, simply set it before calling the code.

Reply:
Internally, I wanted the bootstrapping / random sampling to be consistent across runs, so the result is the same no matter what. If I only set a seed, then the following give difff results:

```
np.random.seed(12345)
stat, pvalue =ccit(x,y,z, )
```
vs.
```
np.random.seed(12345)
do_something_before()
stat, pvalue =ccit(x,y,z, )
```

https://stackoverflow.com/questions/56119505/what-is-the-difference-between-np-random-seed-vs-random-state

Reply:
But if you run multiple random functions in sequence then you would have to re seed the global seed before every function to make it 100% reproducible? My impression was that setting a global seed in general is undesirable (Tho not sure of the detailed intricacies besides this issue).

I just adopted a sklearn pattern here https://scikit-learn.org/stable/developers/utilities.html

Reply:
Ah okay I think I found why: https://stackoverflow.com/questions/59105921/why-is-numpy-random-seed-not-remaining-fixed-but-randomstate-is-when-run-in-para

IIUC the global seed is fine for now but we would in principle add joblib to every function. At that point I think explicitly calling np.random.* seems to have some issues that numpy warns against.

Reply:
We also use global seeds with joblib. Its correct that this requires a careful implementation of the parallelization, but take for instance a look at:
- https://github.com/py-why/dowhy/blob/main/dowhy/gcm/independence_test/kernel.py#L111
- https://github.com/py-why/dowhy/blob/main/dowhy/gcm/independence_test/kernel.py#L487
- https://github.com/py-why/dowhy/blob/main/dowhy/gcm/shapley.py#L219

where we solve this simply by generating a list of random seeds (based on the current global random state) and pass these seeds into the parallel processes. Since the generated seeds are deterministic with a certain global seed, the parallel runs are deterministic as well. This allows you to have also a deterministic behavior when using joblib.